### PR TITLE
[SPARK-48898][SQL] Add Variant shredding functions

### PR DESCRIPTION
**Pull request by cashmand**
### What changes were proposed in this pull request?

This is a first step towards adding Variant shredding support for the Parquet writer. It adds functionality to convert a Variant value to an InternalRow that matches the current shredding spec in https://github.com/apache/parquet-format/pull/461.

Once this merges, the next step will be to set up the Parquet writer to accept a shredding schema, and write these InternalRow values to Parquet instead of the raw Variant binary.

### Why are the changes needed?

First step towards adding support for shredding, which can improve Variant performance (and will be important for functionality on the read side once other tools begin writing shredded Variant columns to Parquet).

### Does this PR introduce _any_ user-facing change?

No, none of this code is currently called outside of the added tests.

### How was this patch tested?

Unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: common/variant/src/main/java/org/apache/spark/types/variant/VariantShreddingWriter.java [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+   * @return The scalar value, or null if the cast is not valid.
+   */
+  private static Object tryTypedShred(
+          Variant v,
+          VariantUtil.Type variantType,
+          VariantSchema.ScalarType targetType,
+          ShreddedResultBuilder builder) {
+    switch (variantType) {
+      case LONG:
+        if (targetType instanceof VariantSchema.IntegralType) {
```
Comment (chenhao-db) [inline common/variant/src/main/java/org/apache/spark/types/variant/VariantShreddingWriter.java:160]: I think we can use the new Java grammar:
```
 if (targetType instanceof VariantSchema.IntegralType integralType) {
```

Comment (cashmand) [reply common/variant/src/main/java/org/apache/spark/types/variant/VariantShreddingWriter.java]: Thanks, TIL.

File: common/variant/src/main/java/org/apache/spark/types/variant/VariantShreddingWriter.java [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+    } else if (schema.objectSchema != null && variantType == VariantUtil.Type.OBJECT) {
+      Map<String, VariantSchema.ObjectField> objectSchema = schema.objectSchema;
+      int maxSize = Math.min(v.objectSize(), objectSchema.size());
+      ShreddedResult[] shreddedValues = new ShreddedResult[objectSchema.size()];
+
+      // Create a variantBuilder for any mismatched fields.
+      VariantBuilder variantBuilder = new VariantBuilder(false);
+      ArrayList<VariantBuilder.FieldEntry> fieldEntries = new ArrayList<>();
+      // Keep track of which schema fields we actually found in the Variant value.
+      Set<String> presentKeys = new HashSet<String>();
```
[the remark this answers is not in the document]
Comment (cashmand) [reply common/variant/src/main/java/org/apache/spark/types/variant/VariantShreddingWriter.java]: Okay, I was on the fence about how much of an effort to put into efficiency here.

I modified the interface so that the shredding schema constructor takes an array of name-schema pairs. I like this better than the map in any case, since my code was assuming that the indices in the map were dense and corresponded to the array being returned in the shredded result, and this makes that much clearer. I now construct the name->field mapping in the constructor. Let me know what you think.

I didn't bother with trying to create an ordered list of fields, I just iterate over the array a second time to fill in nulls if not all fields matched. I think it could be a bit more efficient to sort the list and step through the list and the Variant object together, but it adds more complexity and branching to the loop, and I'm not sure that it would be significantly faster than the simple second iteration over the loop to fill in null values. Let me know if you had something different in mind.

**Comment by Zouxxyy** [comment]
Exciting to see shredding being pushed forward! So If I understand correctly, the shredding write chain may be like this:
Get the expected shredded schema (DataType) through some ways (sampling or justs user defined?) -> parquet writer accepts the variant + shredded schema -> cast to shredded InternalRow -> write to parquet file (the actual col type is the group type corresponding to the shredded dataType).

**Comment by cashmand** [comment]
> So If I understand correctly, the shredding write chain may be like this:
Get the expected shredded schema (DataType) through some ways (sampling or justs user defined?) -> parquet writer accepts the variant + shredded schema -> cast to shredded InternalRow -> write to parquet file (the actual col type is the group type corresponding to the shredded dataType).

Yes, exactly. For initial implementation and testing, I plan to set the schema explicitly. I think sampling makes sense as a better user experience in the long term, but needs some thought about the best way to implement it.

> From this perspective, consider the integration with the lake format, lake format generally has its own reader or writer. I feel like that the lake format may still receive the raw variant data, and the same shredding logic must be implemented in the format's wrtier. If you have any other ideas on it, I'd love to hear your perspective, thanks!

What do you mean by "lake format"? Are you referring to formats like Iceberg or Delta? I made an effort in this PR to keep the shredding logic in common/variant, and created an interface (ShreddedResult and ShreddedResultBuilder) that Spark implements to construct the InternalRow that it uses in its Parquet writer. The intent is that other writers could implement the interface to match their data types, and still reuse the same code for the shredding logic. Eventually, we can separate common/variant into its own Java library (and maybe move it to the Parquet project, which is where the Variant spec is moving to), to make it easier for other JVM-based writers to use the Variant implementation outside of Spark.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+  /**
+   * Given an expected schema of a Variant value, returns a suitable schema for shredding, by
+   * inserting appropriate intermediate value/typed_value fields at each level.
+   * For example, to represent the JSON {"a": 1, "b": "hello"},
+   * the schema struct<a: int, b: string> could be passed into this function, and it would return
+   * the shredding scheme:
+   * struct<
+   *  metadata: binary,
+   *  value: binary,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala:44]: is the value duplicated? In addition to the typed_value, we also store the raw value for each typed field, right?

**Review by gene-db — COMMENTED** [review]
@cashmand Thanks for this feature! I left several questions/comments.

File: common/variant/src/main/java/org/apache/spark/types/variant/VariantSchema.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+  public final int typedIdx;
+  public final int variantIdx;
+  public final int topLevelMetadataIdx;
+  public final int numFields;
+
+  public final ScalarType scalarSchema;
+  public final ObjectField[] objectSchema;
+  // Map for fast lookup of object fields by name.
+  public final Map<String, Integer> objectSchemaMap;
```
[the remark this answers is not in the document]
Comment (cashmand) [reply common/variant/src/main/java/org/apache/spark/types/variant/VariantSchema.java:117]: Yes.

File: common/variant/src/main/java/org/apache/spark/types/variant/VariantShreddingWriter.java [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+      for (int i = 0; i < size; ++i) {
+        ShreddedResult shreddedArray = castShredded(v.getElementAtIndex(i), elementSchema, builder);
+        array[i] = shreddedArray;
+      }
+      result.addArray(schema, array);
+    } else if (schema.objectSchema != null && variantType == VariantUtil.Type.OBJECT) {
+      VariantSchema.ObjectField[] objectSchema = schema.objectSchema;
+      ShreddedResult[] shreddedValues = new ShreddedResult[objectSchema.length];
+
+      // Create a variantBuilder for any mismatched fields.
```
[the remark this answers is not in the document]
Comment (cashmand) [reply common/variant/src/main/java/org/apache/spark/types/variant/VariantShreddingWriter.java]: The latter. I'll reword it to clarify.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+case object SparkShreddingUtils {
+  def buildVariantSchema(schema: DataType): VariantSchema = {
+    schema match {
+      case s: StructType => buildVariantSchema(s, topLevel = true)
+      case _ => throw QueryCompilationErrors.invalidVariantSchema(schema)
+    }
+  }
+
+  /**
+   * Given an expected schema of a Variant value, returns a suitable schema for shredding, by
```
[the remark this answers is not in the document]
Comment (cashmand) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala:41]: `VariantSchema` is a concept in the `common/variant` code, and represents a shredding schema (possibly one that is nested as an array element or object field in a larger shredding schema).

`StructType` is a Spark type. It happens to be used as the Spark implementation to represent both `VariantSchema` and a shredded object in such a schema.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+   *   b: struct<typed_value: string, value: binary>>>
+   *
+   */
+  def variantShreddingSchema(dataType: DataType, isTopLevel: Boolean = true): StructType = {
+    val fields = dataType match {
+      case ArrayType(elementType, containsNull) =>
+        val arrayShreddingSchema =
+          ArrayType(variantShreddingSchema(elementType, false), containsNull)
+        Seq(
+          StructField("value", BinaryType, nullable = true),
```
[the remark this answers is not in the document]
Comment (cashmand) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala]: Okay, I'll put it in VariantaSchema, since it's part of the spec.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+          StructField("typed_value", dataType, nullable = true)
+        )
+      case _ =>
+        // No other types have a corresponding shreddings schema.
+        throw QueryCompilationErrors.invalidVariantSchema(dataType)
+    }
+
+    if (isTopLevel) {
+      StructType(StructField("metadata", BinaryType, nullable = true) +: fields)
```
[the remark this answers is not in the document]
Comment (cashmand) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala]: You're right, I'll change it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        case "metadata" =>
+          if (topLevelMetadataIdx != -1 || f.dataType != BinaryType) {
+            throw QueryCompilationErrors.invalidVariantSchema(schema)
+          }
+          topLevelMetadataIdx = i
+        case _ => throw QueryCompilationErrors.invalidVariantSchema(schema)
+      }
+    }
+
+    if (topLevel != (topLevelMetadataIdx >= 0)) {
```
[the remark this answers is not in the document]
Comment (cashmand) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala:159]: Yes, it's just for validation. In the long term, I wasn't sure if we'd necessarily use `variantShreddingSchema`, or if data sources would directly provide the full shredding schema, so I think it's good to validate here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+    }
+    new VariantSchema(typedIdx, variantIdx, topLevelMetadataIdx, schema.fields.length,
+      scalarSchema, objectSchema, arraySchema)
+  }
+
+  class SparkShreddedResult(schema: VariantSchema) extends VariantShreddingWriter.ShreddedResult {
+    // Result is stored as an InternalRow.
+    val row = new GenericInternalRow(schema.numFields)
+
+    override def addArray(schema: VariantSchema,
```
[the remark this answers is not in the document]
Comment (cashmand) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala]: Oh, that's a good point. I don't think it's necessary to pass the schema into the methods. I'll try to remove it.

**Review by cashmand — COMMENTED** [review]
Thanks for the review, @gene-db! I updated the PR based on your comments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (cashmand) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/SparkShreddingUtils.scala]: Never mind, I ended up just putting it in `SparkShreddingUtils`. It doesn't seem to like using Java constants in cast matches, and also adds extra verbosity.

File: common/variant/src/main/java/org/apache/spark/types/variant/VariantSchema.java [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+  public static final class DateType extends ScalarType {
+  }
+
+  public static final class TimestampType extends ScalarType {
+  }
+
+  public static final class TimestampNTZType extends ScalarType {
+  }
+
+  // The index of the typed_value, value, and metadata fields in the schema, respectively. I given
```
Comment (gene-db) [inline common/variant/src/main/java/org/apache/spark/types/variant/VariantSchema.java:102]:
```suggestion
  // The index of the typed_value, value, and metadata fields in the schema, respectively. If given
```

**Review by gene-db — APPROVED** [review]
@cashmand Thanks for these functions!

LGTM

**Comment by cloud-fan** [comment]
The docker test failure is unrelated, thanks, merging to master!

**Comment by adriangb** [comment]
Hi folks, sorry to bother you, could some of you who are familiar with these bits of Spark please post example files to https://github.com/apache/parquet-testing so that non-Spark devs can start working on the format in other implementations? Thanks!
